### PR TITLE
Add support for partition and hidden columns to OrcSelectivePageSource

### PR DESCRIPTION
**Pull request by mbasmanova**
Depends on #13066

### CODE REVIEW

**Comment by oerling** [comment]
Does the OrcSelectiveRecordReader  know the column Types?

 

The list of Type from HivePageSource is definitely needed for structured types, e.g. there will be a map key column that is expected to be exactly int or long or short or byte and you do not know which based on the StreamReader alone.  

 

OrcType tells you what the split contains. The Type from the  Operator tells you what kind of Block to make.

 

I found that the column Types from the HivePageSource and from the OrcPageSource can differ in a split dependent manner. The record reader should know both in order to figure out coercions. If it knows these it has no need for coercers from the level above and you do not need to fill these in for splits at the split source for the filter-only columns. Coercions on prefilled should not occur because these are partitioning columns and have a table wide type. 

 

Coercions do not matter for absent columns because these are always null. So maybe you can assert they are null, for the eventuality that these got a non-null default someday.

 

For reference you can check out CoercedFilter in Filters.java on test-prod2, as well as references thereto.

 

Coercion is very rare and does not have to be particularly efficient but needs to exist.

**Comment by oerling** [comment]
Please ignore the previous, I think all the information that is needed is present.

 

Is this intended to work in the case of missing columns?

 

I try the schema evolution test table from Aria and I get an index out of bounds exception in creating the stream readers.

 

I think it would be good to cover these here, for example add a prefilled column with a null value for the case of a missing regular column. The sane mechanism will work for these. Having two kinds of prefilled columns can be confusing.

**Comment by mbasmanova** [comment]
@oerling Orri, thank you for review.

> Is this intended to work in the case of missing columns?

This PR doesn't add support for coercions or missing columns. This functionality will appear in future PRs.

**Comment by mbasmanova** [comment]
@highker 

> "Add support for partition and hidden columns to OrcSelectivePageSource" LGTM

James, thank you for review. My reading of the above suggests that you'll approve this PR once the dependent PR #13066 is approved. Is this correct?

**Comment by mbasmanova** [comment]
@tdcmeehan @highker Rebased onto latest master. James, Tim, would you take a look?

**Comment by mbasmanova** [comment]
@highker @tdcmeehan Thank you, James and Tim.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java [hunk 1]
```diff
@@ -254,9 +281,14 @@ public Page getNextPage()
...
             int columnIndex = outputColumns.get(i);
-            Block block = getStreamReaders()[columnIndex].getBlock(positionsToRead, positionCount);
-            updateMaxCombinedBytesPerRow(columnIndex, block);
-            blocks[i] = block;
+            if (constantValues[columnIndex] != null) {
+                blocks[i] = RunLengthEncodedBlock.create(columnTypes.get(columnIndex), constantValues[columnIndex], batchSize);
+            }
+            else {
+                Block block = getStreamReader(columnIndex).getBlock(positionsToRead, positionCount);
+                updateMaxCombinedBytesPerRow(columnIndex, block);
```
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:289]: Sorry for commenting on a very old PR.  should the columnIndex here be hiveColumnIndices[columnIndex]. I got a crash with the following call stack.

```
com.facebook.presto.spi.PrestoException: Failed to read ORC file: ws://wsXXXXXXXX
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:106)
	at com.facebook.presto.operator.ScanFilterAndProjectOperator.processPageSource(ScanFilterAndProjectOperator.java:277)
	at com.facebook.presto.operator.ScanFilterAndProjectOperator.getOutput(ScanFilterAndProjectOperator.java:242)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:418)
	at com.facebook.presto.operator.Driver.lambda$processFor$9(Driver.java:301)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:722)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:294)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1077)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:599)
	at com.facebook.presto.$gen.Presto_0_257_df401fd__0_257____20210728_224554_1.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 5 out of bounds for length 5
	at com.facebook.presto.orc.AbstractOrcRecordReader.updateMaxCombinedBytesPerRow(AbstractOrcRecordReader.java:397)
	at com.facebook.presto.orc.OrcSelectiveRecordReader.getNextPage(OrcSelectiveRecordReader.java:725)
	at com.facebook.presto.hive.orc.OrcSelectivePageSource.getNextPage(OrcSelectivePageSource.java:86)
	... 13 more
```

**Review by arunthirupathi — COMMENTED** [review]
If the comment looks ok, I can send a PR for it.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java (hunk 1 above)
Comment (mbasmanova) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:289]: @arunthirupathi Indeed. Looks like we should use hiveColumnIndices[columnIndex] here (just like we do in getStreamReader:

```
    private SelectiveStreamReader getStreamReader(int columnIndex)
    {
        return getStreamReaders()[hiveColumnIndices[columnIndex]];
    }
```

Do you think you can create a test case to reproduce this issue?

Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:289]: I will create a PR with the test to repro this.

**Pull request merged**